### ISSUE TITLE
Overlapping masks issue

### ISSUE THREAD

**Issue by lwdevelop191** [roboflow/rf-detr#976]
Hello,

I have been doing some test with the segmentation model and I am running into an issue.
I have some cases where and object will be partially in front of another one and right now the model is giving me a mask for the whole item that is behind the forward one, causing two masks that have overlap.
Is there anything I can do about this? When testing with some YOLO models I never saw this issue using the same dataset.

**Comment by isaacrob** [comment]
Are you annotating them with overlapping or disjoint masks?

**Comment by lwdevelop191** [comment]
Disjoint masks. I have been training using the YOLO annotation format instead of COCO, not sure if this has any effect.

**Comment by isaacrob** [comment]
If training with disjoint masks then the model should converge to predicting disjoint masks. Can you post your training parameters and an example of where it's failing?

**Comment by lwdevelop191** [comment]
Here an example with my training code

<img width="336" height="239" alt="Image" src="https://github.com/user-attachments/assets/90553ab7-4682-4f57-95ef-2843d6189d0d" />

```python
from rfdetr import RFDETRSegXLarge

TRAIN_RES = 624 

boxes_aug = {
    "Resize": {"height": TRAIN_RES, "width": TRAIN_RES, "p": 1.0},
    "HorizontalFlip": {"p": 0.5},
    
    # Exposure (+- 10%)
    "RandomBrightnessContrast": {
        "brightness_limit": 0.1, 
        "contrast_limit": 0.1, 
        "p": 0.5
    },
    
    # Saturation & Hue
    "HueSaturationValue": {
        "hue_shift_limit": 10, 
        "sat_shift_limit": 20, 
        "val_shift_limit": 10, 
        "p": 0.5
    },
    
    # Random Crop & Zoom met de centrale variabele
#    "RandomSizedBBoxSafeCrop": {
#        "height": TRAIN_RES, 
#        "width": TRAIN_RES, 
#        "erosion_rate": 0.0, 
#        "p": 0.2
#    }
}

model = RFDETRSegXLarge()

model.train(
    dataset_dir=dataset.location, 
    epochs=200, 
    batch_size=4, 
    grad_accum_steps=4,
#    resume="output/last.ckpt",
    progress_bar=True,
    num_workers=6,
    maxDets = 500,
    persistent_workers=True,
    early_stopping=True,
    early_stopping_patience=10,
    early_stopping_min_delta=0.002,
    early_stopping_use_ema=True,
    aug_config=boxes_aug,
)

**Comment by isaacrob** [comment]
Wait I don't understand what's happening in that image 😅 can you send the clean image? And a description of what you're trying to do

**Comment by lwdevelop191** [comment]
It are layers with cardboard boxes. You can see that the masks of one in the background is going over a box in the front.

**Comment by Borda** [comment]
> Here an example with my training code
> 
> <img alt="Image" width="336" height="239" src="https://private-user-images.githubusercontent.com/136300869/579965441-90553ab7-4682-4f57-95ef-2843d6189d0d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzY2ODIzODksIm5iZiI6MTc3NjY4MjA4OSwicGF0aCI6Ii8xMzYzMDA4NjkvNTc5OTY1NDQxLTkwNTUzYWI3LTQ2ODItNGY1Ny05NWVmLTI4NDNkNjE4OWQwZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwNDIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDQyMFQxMDQ4MDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wYzBlN2IwYjMyNzMxMTIzMDA3Y2I4YTVhZDdjODA3MmFlNTNmNTVhOGVkNzEzZDVkYTVkZTQ2ODllMTdjNWVjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ.aWrN7Ej7PN32kpcwfBouaVN7EOfLEGCRYeDBtkZN6O8">

Based on the picture, there is no overlap on the pixel level... however, you can say that there is an overlap on the physical world level, as you assume the model would do some reasoning and extrapolation to undestand that the paper has a rectangular shape...

**Comment by lwdevelop191** [comment]
For comparison this is how the YOLO model handles it.

<img width="229" height="233" alt="Image" src="https://github.com/user-attachments/assets/f2f9df2b-38eb-4b2c-b935-911f2a532c21" />

**Comment by isaacrob** [comment]
Super interesting! Are there images like that in the train set too?

**Comment by lwdevelop191** [comment]
Yes the train set has some images like this. Always annotated in the way that the YOLO model shows.

**Comment by isaacrob** [comment]
Would you consider not setting the maxDets argument? That's not compatible with the pretrained weights. Are there images where there are 500 boxes present?

**Comment by lwdevelop191** [comment]
The max amount will be around 200-250. That is why I went for the XLarge model. I can remove the maxDets is that is better.

**Comment by zhaoshuo1223** [comment]
Is this dataset available for sharing? I would like to reproduce the results. Thank you very much.

**Comment by isaacrob** [comment]
@lwdevelop191 yeah I would encourage you to just try it without any modifications and see if it works in that case

**Comment by lwdevelop191** [comment]
@isaacrob The first model I trained was without the maxDets. This model shows the same behavior.

**Comment by isaacrob** [comment]
Oh got it. Super interesting! First real 'failure' mode I remember seeing :) 

For how many epochs did it train?

**Comment by lwdevelop191** [comment]
I think around 30 epochs, then the early stop kicked in. I also notice that if I run the model with the same input I sometimes get different results.  Sometimes it misses some boxes that the next run it sees without a problem and quite a high score.

**Comment by isaacrob** [comment]
Oh that shouldn't happen. The results should always be the same between forward passes with the same weights. How are you running it?

**Comment by lwdevelop191** [comment]
I am running the model using Openvino. I think the issue was the fp16 rounding on the GPU. If I force fp32 it seems to be consistent.

**Comment by isaacrob** [comment]
Oh interesting. Fp16 can cause issues if the runtime doesn't handle it correctly. Is the issue still present in fp32?

**Comment by lwdevelop191** [comment]
In fp16 it sometimes misses some objects even if I use the same picture a couple of times. One time it sees it a few times not. fp32 the results seem stable.

**Comment by isaacrob** [comment]
I'm wondering if the above masking issue that started this thread happens in fp32 or if it's only in fp16?

**Comment by lwdevelop191** [comment]
Also happens in fp32. Is always like this. It is never the correct way

**Comment by isaacrob** [comment]
Oh I see, I misunderstood your earlier comment. 

Would you mind trying training in 1.4.1?
